python/cugraph/structure/new_number_map.py

```diff
@@ -0,0 +1,275 @@
+# Copyright (c) 2021, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from dask.distributed import wait, default_client
+from cugraph.dask.common.input_utils import get_distributed_data
+from cugraph.structure import renumber_wrapper as c_renumber
+import cugraph.comms as Comms
+import dask_cudf
+import numpy as np
+import cudf
+
+
+def call_renumber(sID,
+                  data,
+                  num_verts,
+                  num_edges,
+                  is_mnmg):
+    wid = Comms.get_worker_id(sID)
+    handle = Comms.get_handle(sID)
+    return c_renumber.renumber(data[0],
+                               num_verts,
+                               num_edges,
+                               wid,
+                               handle,
+                               is_mnmg)
+
+class NumberMap:
+
+    class SingleGPU:
+        def __init__(self, df, src_col_names, dst_col_names, id_type,
+                     store_transposed):
+            self.col_names = NumberMap.compute_vals(src_col_names)
+            self.df = cudf.DataFrame()
+            self.id_type = id_type
+            self.store_transposed = store_transposed
+            self.numbered = False
+
+        def to_internal_vertex_id(self, df, col_names):
+            tmp_df = df[col_names].rename(
+                columns=dict(zip(col_names, self.col_names)), copy=False
+            )
+            index_name = NumberMap.generate_unused_column_name(df.columns)
+            tmp_df[index_name] = tmp_df.index
+            return (
+                self.df.merge(tmp_df, on=self.col_names, how="right")
+                .sort_values(index_name)
+                .drop(columns=[index_name])
+                .reset_index()["id"]
+            )
+
+        def from_internal_vertex_id(
+            self, df, internal_column_name, external_column_names
+        ):
+            tmp_df = self.df.merge(
```

Comment:
Have you checked the performance of this?

From internal to external vertex ID, it will simply be external_vertex_ID = renumber_map[internal_vertex_ID] in SG. Should be very fast. I guess this merge can be much slower.

Reply:
This takes into account multi column unrenumbering

Reply:
I see, and isn't this unnecessarily expensive in single column cases? We may take a cheaper approach if the original vertex IDs are single-column?

Reply:
This can be addressed as part of optimization in subsequent PR